packages/core/source-map/src/SourceMap.js

```diff
@@ -0,0 +1,441 @@
+    columnOffset: number = 0
+  ) {
+    consumer.eachMapping(mapping => {
+      this.addConsumerMapping(mapping, lineOffset, columnOffset);
+
+      if (!this.sources.has(mapping.source)) {
+        this.sources.set(
+          mapping.source,
+          consumer.sourceContentFor(mapping.source, true)
+        );
+      }
+    });
+
+    consumer.destroy();
+
+    return this;
+  }
+
+  async addMap(
+    map: RawMapInputType | SourceMap,
+    lineOffset: number = 0,
+    columnOffset: number = 0
+  ) {
+    if (typeof map === 'string' || typeof map.mappings === 'string') {
+      let consumer = await this.getConsumer(map);
+
+      return this._addConsumerMap(consumer, lineOffset, columnOffset);
+    } else if (map.mappings && map.sources) {
+      if (!map.eachMapping) {
+        // $FlowFixMe
+        map = new SourceMap(map.mappings, map.sources);
+      }
+
+      // TODO: Not sure if this is even necessary?
+      if (!(map instanceof SourceMap)) {
+        throw new Error(
+          'Let me know if this threw, Flow.js said it might happen ~Jasper'
+        );
+      }
+
+      return this._addSourceMap(map, lineOffset, columnOffset);
+    } else {
+      throw new Error('Could not merge sourcemaps, input is of unknown kind');
+    }
+  }
+
+  addMapping(
+    mapping: MappingType,
+    lineOffset: number = 0,
+    columnOffset: number = 0
+  ) {
+    this.mappings.push({
+      source: mapping.source,
+      name: mapping.name,
+      original: mapping.original,
+      generated: {
+        line: mapping.generated.line + lineOffset,
+        column: mapping.generated.column + columnOffset
+      }
+    });
+  }
+
+  addConsumerMapping(
+    mapping: ConsumerMappingItemType,
+    lineOffset: number = 0,
+    columnOffset: number = 0
+  ) {
+    let original = null;
+    if (
+      typeof mapping.originalLine === 'number' &&
+      mapping.originalLine > 0 &&
+      typeof mapping.originalColumn === 'number'
+    ) {
+      original = {
+        line: mapping.originalLine,
+        column: mapping.originalColumn
+      };
+    }
+
+    this.mappings.push({
+      source: original ? mapping.source : null,
+      name: mapping.name,
+      original,
+      generated: {
+        line: mapping.generatedLine + lineOffset,
+        column: mapping.generatedColumn + columnOffset
+      }
+    });
+  }
+
+  eachMapping(callback: (mapping: MappingType) => any) {
+    this.mappings.forEach(callback);
+  }
+
+  generateEmptyMap(sourceName: string, sourceContent: string) {
+    this.sources.set(sourceName, sourceContent);
+
+    this.lineCount = lineCounter(sourceContent);
+    for (let line = 1; line < this.lineCount + 1; line++) {
+      this.addMapping({
+        source: sourceName,
+        original: {
+          line: line,
+          column: 0
+        },
+        generated: {
+          line: line,
+          column: 0
+        }
+      });
+    }
+
+    return this;
+  }
+
+  async extendSourceMap(extension: SourceMap | RawMapInputType) {
```

Comment:
I've changed this from `async extendSourceMap(original, extension) {` into `async extendSourceMap(extension) {` as we already use it kind of this way in Parcel 1.
In Parcel 2 we will probably pass through and cache instances of the parcel sourcemap so this seemed like it might be faster as well.